pkg/server/sandbox_status.go

```diff
@@ -82,62 +86,70 @@ func (c *criContainerdService) PodSandboxStatus(ctx context.Context, r *runtime.
+	Pid         uint32                    `json:"pid"`
+	Status      string                    `json:"processStatus"`
+	NetNS       string                    `json:"netNamespace"`
+	NetNSClosed bool                      `json:"netNamespaceClosed"`
+	Image       string                    `json:"image"`
+	SnapshotKey string                    `json:"snapshotKey"`
+	Snapshotter string                    `json:"snapshotter"`
+	Config      *runtime.PodSandboxConfig `json:"config"`
+	RuntimeSpec *runtimespec.Spec         `json:"runtimeSpec"`
 }
 
-// toCRISandboxContainerInfo converts internal container object information to CRI sandbox container status response info map.
-func toCRISandboxContainerInfo(ctx context.Context, container containerd.Container, verbose bool) (map[string]string, error) {
+// toCRISandboxInfo converts internal container object information to CRI sandbox status response info map.
+func toCRISandboxInfo(ctx context.Context, sandbox sandboxstore.Sandbox,
+	pid uint32, processStatus containerd.ProcessStatus, verbose bool) (map[string]string, error) {
 	if !verbose {
 		return nil, nil
 	}
 
-	task, err := container.Task(ctx, nil)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get sandbox container task: %v", err)
-	}
-	status, err := task.Status(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get sandbox container task status: %v", err)
+	si := &sandboxInfo{
+		Pid:    pid,
+		Status: string(processStatus),
+		NetNS:  sandbox.NetNSPath,
+		Config: sandbox.Config,
 	}
 
-	info := make(map[string]string)
-	pid := task.Pid()
+	if si.Status == "" {
+		// If processStatus is empty, it means that the task is deleted. Apply "deleted"
+		// status which does not exist in containerd.
+		si.Status = "deleted"
+	}
 
-	oldSpec, err := container.Spec(ctx)
-	if err != nil {
-		glog.Errorf("Failed to get sandbox container runtime spec: %v", err)
+	if sandbox.NetNSPath != "" {
+		// Add network closed information if sandbox is not using host network.
+		si.NetNSClosed = (sandbox.NetNS == nil || sandbox.NetNS.Closed())
 	}
 
-	var snapshotkey, snapshotter string
-	ctrInfo, err := container.Info(ctx)
+	container := sandbox.Container
+
+	spec, err := container.Spec(ctx)
 	if err == nil {
-		snapshotkey = ctrInfo.SnapshotKey
-		snapshotter = ctrInfo.Snapshotter
+		si.RuntimeSpec = spec
 	} else {
-		glog.Errorf("Failed to get target shapshot info: %v", err)
-	}
-
-	si := &sandboxInfo{
-		Pid:         pid,
-		State:       string(status.Status),
-		SnapshotKey: snapshotkey,
-		Snapshotter: snapshotter,
-		RuntimeSpec: oldSpec,
+		glog.Errorf("Failed to get sandbox container %q runtime spec: %v", sandbox.ID, err)
 	}
 
-	m, err := json.Marshal(si)
+	ctrInfo, err := container.Info(ctx)
 	if err == nil {
-		info["info"] = string(m)
+		// Do not use config.SandboxImage because the configuration might
+		// be changed during restart. It may not reflect the actual image
+		// used by the sandbox container.
+		si.Image = ctrInfo.Image
```

Comment:
all good to update if these are not nil?

Reply:
What do you mean?

Reply:
Should we do a `if ctrInfo.Image != ""` check?

Reply:
Seems not. If it is empty, then just assign an empty value. No difference.

Reply:
I might've been confused by the comment. Not sure why we would update with nil here, if that is as reasonable thing to report/do. Just felt off :-)

Reply:
I just mean the image in container info is the source of truth. We should use it whatever it is. :)